### ISSUE TITLE
Fatal error: Call to a member function push() on a non-object in phar

### ISSUE THREAD

**Issue by nicksavov** [OSTraining/AllediaBuilder#15]
When trying `phing build` for OSToolbar's development branch (https://github.com/OSTraining/ostoolbar), I'm getting the following error in the command line:

> Fatal error: Call to a member function push() on a non-object in phar:///Users/Savov/Documents/phing/phing-latest.phar/classes/phing/contrib/DocBlox/Parallel/Manager.php on line 237

**Comment by billtomczak** [comment]
Looks like there might be a problem with your phing setup or version.

**Comment by andergmartins** [comment]
Agree. Just tested and it worked fine. Do you have the property file set inside OSToolbar folder? On the other repos we have a file called **build.dist.properties** or **build.properties.dist**. Duplicate it and commit into OSToolbar repo.
So inside OSToolbar, confirm if you have a **build.properties** file like this:

    builder.path=/home/vagrant/Projects/AllediaBuilder

    project.AllediaFramework.path=/home/vagrant/Projects/AllediaFramework
    project.AllediaInstaller.path=/home/vagrant/Projects/AllediaInstaller
    project.AllediaFreeDefaultFiles.path=/home/vagrant/Projects/AllediaFreeDefaultFiles
    project.OSSystem.path=/home/vagrant/Projects/OSSystem

**Comment by nicksavov** [comment]
I'm running the latest version of phing and I have the build.properties file, however, I don't have the "project.AllediaFreeDefaultFiles.path" in there. Is that necessary?

**Comment by nicksavov** [comment]
Relevant code:
https://www.phing.info/trac/browser/classes/phing/contrib/DocBlox/Parallel/Manager.php#L237

**Comment by billtomczak** [comment]
@andergmartins should have the last word on this, but I do not have that line in the build.properties file and it works okay for me. I'm guessing that's either an obsolete property or only used when the package-license is 'pro' (as opposed to 'free').

**Comment by billtomczak** [comment]
Ah! Your message came in after I responded. :)

Your OP referred to trying to build OSToolbar but that code is looking for OSCampus?

**Comment by andergmartins** [comment]
We only need that line and the AllediaFreeDefaultFiles cloned for Alledia Extensions which are using the alledia footer with the link to JED. That repo was the way I found to keep only one instance of the code required to show the footer and store any file that should be common among all Alledia Extensions. 
That should only be required if the manifest file is referencing that repo in an "include" tag (child of Alledia's tag). Since OSToolbar is going to Alledia (and SR as well), I think we should use that, and add the common footer too. BUT, that is not required for the first version, or to just publish in the JED. If that is the case, remove it from the manifest and you will not need to use that repo, nor that line on the build.properties file.
But after the first version, I think we should update it and SimpleRenew in order to display the Alledia footer

**Comment by billtomczak** [comment]
Cool. Thanks for that explanation @andergmartins

**Comment by andergmartins** [comment]
:) yw!

**Comment by nicksavov** [comment]
@billtomczak , yeah, I'm trying both. 

For OSCampus, it's:
Fatal error: Call to a member function push() on a non-object in phar:///Users/User/Documents/phing/phing-latest.phar/classes/phing/contrib/DocBlox/Parallel/Manager.php on line 237

If I change the contents of OSCampus/src/oscampus.xml to just:
> < extension>< /extension>

I still get the same error. Note, I added an extra space after the opening < above, so that github would accept the markup.

However, if I change the contents of the file to:
>  a

Then I get:

> [PHP Error] simplexml_load_file(): /Users/User/Documents/OSTraining/Our-Extensions/git/github/OSCampus/src/oscampus.xml:1: parser error : Start tag expected, '<' not found [line 221 of phar:///Users/User/Documents/phing/phing-latest.phar/classes/phing/tasks/ext/XmlPropertyTask.php]
[PHP Error] simplexml_load_file(): a [line 221 of phar:///Users/User/Documents/phing/phing-latest.phar/classes/phing/tasks/ext/XmlPropertyTask.php]
[PHP Error] simplexml_load_file(): ^ [line 221 of phar:///Users/User/Documents/phing/phing-latest.phar/classes/phing/tasks/ext/XmlPropertyTask.php]

> BUILD FAILED

**Comment by nicksavov** [comment]
I'll try updating to the latest version of Yosemite. Who knows :)

**Comment by andergmartins** [comment]
humm, not sure. If you want, you can provide me remote access later, so I can try debugging it

**Comment by billtomczak** [comment]
@nicksavov, @andergmartins  -- Please note that I updated OSToolbar to include the Free Files dependency.

One thing you should check is that OSSystem builds successfully on its own. Since OSToolbar is dependent on it, you might be getting the error from that repo build.

**Comment by nicksavov** [comment]
Thanks for the hint! I had some build errors in OSSystem's build, which I resolved by adding project.AllediaFreeDefaultFiles.path. Now it just has the same error as OSToolbar and OSCampus' repo :D

**Comment by billtomczak** [comment]
lol!

Sounds like the next thing to check is the path to your local clone of https://github.com/OSTraining/AllediaFreeDefaultFiles

**Comment by nicksavov** [comment]
That looks fine. It's in the same folder as all the others

**Comment by andergmartins** [comment]
Let's run a remote session so I can debug it for you ;)

**Comment by nicksavov** [comment]
Thanks to Anderson's help, https://github.com/OSTraining/AllediaBuilder/commit/7a1468e07ce373454280a522eae382251b643311 fixed it for me :)

**Comment by billtomczak** [comment]
It's possible I wasn't having trouble before because I hadn't updated the builder. Now that I have, I'm seeing the issue as well. Switching to Nick's patch so I can move forward.

**Comment by andergmartins** [comment]
Right, makes sense Bill :)
I think we can maintain both branches, but move the master to master-parallel and rename the no-parallel-bug-fix to master.
For extensions with multiple related extensions like OSMap, I win some seconds in each build what makes me happy :smiley: so I would like to keep the parallel tasks as alternative.

**Comment by billtomczak** [comment]
I'll try to do some sleuthing and see if I can find the problem. I'm not thrilled about maintaining separate branches. Maybe there's a way to make us both happy :)

**Comment by andergmartins** [comment]
;) sounds good. Ok, I see you point about multiple branches. We can have a param in the config file telling us to use or not the parallel task. And a conditional inside the builder code to use or not the parallel.

**Comment by billtomczak** [comment]
That's a reasonable stop-gap until we can figure out what the issue is.

**Comment by billtomczak** [comment]
Finally some time to research this. See: https://www.phing.info/docs/stable/hlhtml/index.html#ParallelTask

Especially:
> Warning
>
> This task is highly experimental, and will only work on *nix machines that have the PHP pcntl extension installed.

So in @nicksavov's case his php installation probably doesn't have the needed extension and in my case, well - it's windows!

**Comment by billtomczak** [comment]
@andergmartins -- see 72530b5404467eb08bd522ecf07a6f2136bd6709 for my proposed solution.

I don't know if there might be a simpler solution than adding those &lt;if&gt; blocks. But it seems to work.

**Comment by andergmartins** [comment]
Hey @billtomczak, sounds good for me. I'm merging it, thanks!

**Comment by andergmartins** [comment]
Merged to master. @nicksavov does the master branch work for you now?
